protocols/mempool-overview.md

```diff
@@ -134,10 +139,134 @@ Committed transactions are not in the mempool:
 
 Finally, we require some progress from the mempool.
 Namely, if a transaction appears at a correct process then eventually it is committed or forever invalid.  
 **INV4** $\forall tx. \forall p \in Correct. \square(tx \in p.mempool \implies \lozenge\square(tx \in p.committed \vee \neg p.ledger.valid(tx)))$
+
+The above invariant ensures that if a transaction enters the mempool, then it eventually leaves it at all the correct processes.
+It requires from the application that the validity of a transaction converges toward some value.
+That is, in every run, there is a height after which $valid(tx)$ always returns the same value.
+This happens in [Ethereum](https://github.com/ethereum/go-ethereum/blob/5c51ef8527c47268628fe9be61522816a7f1b395/light/txpool.go#L401) as a transaction is always valid until a transaction from the same address executes with the same or higher nonce.
+A simpler way to satisfy this requirement is by having $valid(tx)$ deterministic and stateless (e.g., a syntactic check).
+
+**Practical considerations.**
+In practice, as it requires to traverse the whole ledger, INV2 is too expensive.
+Instead, we would like to maintain this only over the last $\alpha$ committed transactions, for some parameter $\alpha$.
+Given a process $p$, we write $p.lcommitted$ the last $\alpha$ committed transactions at $p$.
+Invariant INV2 is replaced with:  
+**INV2a.** $\forall tx. \forall p \in Correct. \square(tx \in p.lcommitted \implies tx \notin p.mempool)$
+
+Another practical concern is with INV3.
+This invariant requires to have a green light from the client application before adding a transaction to the mempool.
+For efficiency, such a validation needs to be made at most $\beta$ times per transaction at each height, for some parameter $\beta$.
+Ideally, $\beta$ equals $1$.
+In practice, $\beta = f(T)$ for some function f of the maximal number of transactions T submitted between two heights.
+Given some transaction $tx$, variable $p.valid[tx]$ tracks the number of times the application was asked at the current height.
+Invariant INV3 is replaced with:  
+**INV3a.** $\forall tx. \forall p \in Correct. \square(tx \in p.mempool \implies p.valid[tx] \in [1, \beta])$
+
+
+## Implementation in CometBFT (as of v0.38.0-alpha.2)
+
+The mempool is implemented in `clist_mempool.go`, in the `CListMempool` data type.
+`CListMempool` uses a single variable for the two mechanisms covered in the previous section.
+Below, we present this approach in detail then establish its correctness.
+
+**Algorithm**
+For starters, we explain at a high-level the logic in `CListMempool`.
+Variables `txs` and `cache` respectively hold the mempool and the validity cache in a FIFO and LRU list.
+We omit the use of `txsMap`, assuming that a transaction is in `txs` iff it is also in `txsMap` (see [this](https://github.com/cometbft/cometbft/pull/890) fix).
+For simplicity, we consider that 
+_(i)_ the mempool is never full, 
```

Comment:
This is a high-level proof. If we have to delve into all the subtelities (and interleavings) of `clist_mempool.go`, this is just an untractable problem.

Hence, we make some simplifications, e.g., txsMap is not there (because anyway, we must now have $tx \in txs iff tx \in txsMap$.